sophia/src/serializer.rs

```diff
@@ -1,217 +1,355 @@
-//! Serializers for standard RDF syntaxes,
-//! and tools for building new serializers.
-//!
-//! # Uniform interface
-//!
-//! Each serializer module defines a `Config` type, that
-//! - implements [`Default`],
-//! - has a `writer` method taking any `io::Write`
-//!   and returning a [`TripleWriter`] or a [`QuadWriter`],
-//! - has a `stringifier` method returning a [`StringSerializer`] or a [`QuadStringifier`].
-//!
-//! Each serializer module also provides two functions `writer` and a `stringifier`,
-//! calling the corresponding methods from the default `Config`.
-//!
-//! [`TripleWriter`] and [`TripleStringifier`] are specializations of [`TripleSink`].
-//! [`QuadWriter`] and [`QuadStringifier`] are specializations of [`QuadSink`].
-//!
-//! [`Default`]: https://doc.rust-lang.org/std/default/trait.Default.html
-//! [`TripleWriter`]: trait.TripleWriter.html
-//! [`TripleStringifier`]: trait.TripleStringifier.html
-//! [`QuadWriter`]: trait.QuadWriter.html
-//! [`QuadStringifier`]: trait.QuadStringifier.html
-//! [`TripleSink`]: ../triple/stream/trait.TripleSink.html
-//! [`QuadSink`]: ../quad/stream/trait.QuadSink.html
-
-use std::convert::Infallible;
+//! API for serializing RDF syntaxes.
+
 use std::io;
 
 use crate::dataset::*;
 use crate::graph::*;
-use crate::quad::{stream::*, *};
+use crate::quad::stream::*;
 use crate::triple::{stream::*, *};
 
 use std::result::Result; // override ::error::Result
 
-#[macro_use]
-pub mod common;
-pub mod nq;
-pub mod nt;
-
-/// An extension of the `TripleSink` trait,
-/// dedicated to serialization to IO streams.
+/// A triple serializer formats triples in some target of type `T`.
+/// The triples may come from a [`TripleSource`] or a [`Graph`].
 ///
-/// See also [`TripleSink`].
-///
-/// [`TripleSink`]: ../triple/stream/trait.TripleSink.html
-///
-pub trait TripleWriter<W: io::Write>: TripleSink<Outcome = ()> + Sized {
-    type Config;
+/// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+/// [`Graph`]: ../graph/trait.Graph.html
+pub trait TripleSerializer<T> {
+    type Sink: TripleSink;
 
-    fn new(write: W, config: Self::Config) -> Self;
+    /// A sink serializing into `target` the triples fed to it.
+    fn sink(&self, target: T) -> Self::Sink;
 
-    /// Serialize the triples from the given source.
-    fn write<TS, T>(&mut self, mut source: TS) -> Result<(), StreamError<TS::Error, Self::Error>>
+    #[inline]
+    /// Convenient shortcut method for serializing a [`TripleSource`].
+    ///
+    /// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+    fn serialize_triples_in<TS>(&self, mut source: TS, target: T) -> TSResult<Self, T, TS::Error>
```

Comment:
_Personal preference_
I prefer not to add `mut` to arguments that are passed by value because you can always do:

```rust
let mut source = source;
```

without any costs. This just makes the API a little bit less verbose.

Also note [*] at line 21 regarding the naming of the mehtod.

Reply:
Actually, the `mut` in front of the parameter name is *not* part of the function signature. It only impacts the body of the function.

If you are not convinced, have a look at [this example] and [that example].

Granted, this is a poor design decision from Rust's authors, because it really *looks* like it is part of the signature. It is misleadin, as your comment shows. So I will do as you suggest, to avoid confusing more people.

That being said, 

[this example]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c00716b589fda4199f3b78220287f7cb
[that example]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=34c919617ad4fdbef287171a9f775d97